tests/test_aggregates.py

```diff
@@ -0,0 +1,253 @@
+
+    def test_invalid_sort_by(self):
+        sort_list = [1, {}, lambda x: x]
+        for sort in sort_list:
+            with pytest.raises(TypeError):
+                aggregator(sort_by=sort)
+
+    def test_invalid_select(self):
+        selectors = ['str', 1, []]
+        for _select in selectors:
+            with pytest.raises(TypeError):
+                aggregator(select=_select)
+
+    def test_invalid_call(self):
+        call_params = ['str', 1, None]
+        for param in call_params:
+            with pytest.raises(TypeError):
+                aggregator()(param)
+
+    def test_call_without_decorator(self):
+        aggregate_instance = aggregator()
+        with pytest.raises(TypeError):
+            aggregate_instance()
+
+    def test_call_with_decorator(self):
+        @aggregator()
+        def test_function(x):
+            return x
+
+        assert hasattr(test_function, '_flow_aggregate')
+
+    def test_groups_of_invalid_num(self):
+        invalid_values = [{}, 'str', -1, -1.5]
+        for invalid_value in invalid_values:
+            with pytest.raises((TypeError, ValueError)):
+                aggregator.groupsof(invalid_value)
+
+    def test_group_by_invalid_key(self):
+        with pytest.raises(TypeError):
+            aggregator.groupby(1)
+
+    def test_groupby_with_valid_type_default_for_Iterable(self):
+        aggregator.groupby(['half', 'even'], default=[-1, -1])
+
+    def test_groupby_with_invalid_type_default_key_for_Iterable(self):
+        with pytest.raises(TypeError):
+            aggregator.groupby(['half', 'even'], default=-1)
+
+    def test_groupby_with_invalid_length_default_key_for_Iterable(self):
+        with pytest.raises(ValueError):
+            aggregator.groupby(['half', 'even'], default=[-1, -1, -1])
+
+    def test_aggregate_hashing(self):
+        # Since we need to store groups on a per aggregate basis in the project,
+        # we need to be sure that the aggregates are hashing and compared correctly.
+        # This test ensures this feature.
+        def helper_default_aggregator_function(jobs):
+            return [jobs]
+
+        def helper_non_default_aggregator_function(jobs):
+            for job in jobs:
+                yield (job,)
+
+        list_of_aggregates = [
+            aggregator(), aggregator(), aggregator(helper_default_aggregator_function),
+            aggregator(helper_non_default_aggregator_function),
+            aggregator(helper_non_default_aggregator_function),
+            aggregator.groupsof(1), aggregator.groupsof(1),
+            aggregator.groupsof(2), aggregator.groupsof(3), aggregator.groupsof(4),
+            aggregator.groupby('half'), aggregator.groupby('half'), aggregator.groupby('even'),
+            aggregator.groupby('half', -1), aggregator.groupby('even', -1),
+            aggregator.groupby(['half', 'even'], default=[-1, -1])
+        ]
+
+        assert len(set(list_of_aggregates)) == 11
+
```

Comment:
Maybe a comment on how this tests hashing works properly would be helpful (i.e. mentioning that you have 11 unique aggregation definitions).